index.js

```diff
@@ -1,71 +1,186 @@
+  mainWindow.webContents.on('did-finish-load', function () {
+    ipc.emit('default-directory', path)
+  })
+
+  // wait for msg from frontend
+  ipc.on('initialize', function (opts) {
+    ipc.emit('initializing')
+    node.init(opts, function (err, res) {
+      if (err) {
+        ipc.emit('initialization-error', err + '')
+      } else {
+        ipc.emit('initialization-complete')
+        ipc.emit('node-status', 'stopped')
+        fs.writeFileSync(IPFS_PATH_FILE, path)
+      }
+    })
+  })
+}
+
+var startTray = function (node) {
+  var poll
+  var statsCache = {}
+
+  ipc.on('request-state', function (event) {
+    ipfsd.version(function (err, res) {
+      if (err) throw err
+      ipc.emit('version', res)
+    })
+
+    if (node.initialized) {
+      ipc.emit('node-status', 'stopped')
+    }
+  })
+
+  ipc.on('start-daemon', function () {
+    ipc.emit('node-status', 'starting')
+    node.startDaemon(function (err, ipfs) {
+      if (err) throw err
+      ipc.emit('node-status', 'running')
+      poll = setInterval(function () { pollStats(ipfs) }, 500)
+
+      Ipfs = ipfs
+    })
+  })
+
+  ipc.on('stop-daemon', function () {
+    ipc.emit('node-status', 'stopping')
+    if (poll) {
+      delete statsCache.peers
+      ipc.emit('stats', statsCache)
+      clearInterval(poll)
+      poll = null
+    }
+
+    node.stopDaemon(function (err) {
+      if (err) throw err
+      ipc.emit('node-status', 'stopped')
+      ipc.emit('stats', statsCache)
+    })
+  })
+
+  ipc.on('open-console', openConsole)
+  ipc.on('open-browser', openBrowser)
+
+  var pollStats = function (ipfs) {
+    ipc.emit('stats', statsCache)
+    ipfs.swarm.peers(function (err, res) {
+      if (err) throw err
+      statsCache.peers = res.Strings.length
+    })
+  }
+}
+
+// main entry point
+ipfsd.local(function (err, node) {
+  if (err) error(err)
+
+  var mb = menubar({
+    dir: __dirname,
+    width: MENU_WIDTH,
+    index: 'file://' + __dirname + '/html/menu.html',
+    show: false,
+    frame: false,
+    type: 'toolbar',
+    icon: TRAY_ICON
+  })
+
+  mb.on('ready', function () {
+    var path
+
+    // find where we saved our ipfs home, or fallback to default
+    try {
+      path = fs.readFileSync(IPFS_PATH_FILE, 'utf-8')
+    } catch (e) {
+      path = process.env.IPFS_PATH ||
+        (process.env.HOME || process.env.USERPROFILE) + '/.ipfs'
+    }
+
+    startTray(node)
+
+    if (!node.initialized) {
+      initialize(path, node)
+    }
+
+    // keep the menu the right size
+    ipc.on('menu-height', function (height) {
+      mb.window.setSize(MENU_WIDTH, height)
+    })
+  })
+})
+
 var apiAddrToUrl = function (apiAddr) {
   var parts = multiaddr(apiAddr).nodeAddress()
   var url = 'http://' + parts.address + ':' + parts.port + WEBUIPATH
   return url
 }
 
-var openWindow = function (ipfs) {
-  if (!mainWindow) {
-    ipfs.config.get('Addresses.API', function (err, res) {
+var openConsole = function () {
+  if (Ipfs) {
```

Comment:
should probably return an error if not, so maybe this func should take an optional callback